pkg/pillar/agentbase/apptracker.go

```diff
@@ -0,0 +1,539 @@
+	UUID      string
+	Name      string
+	Exist     bool
+	Activated LocalTriState
+	State     string
+	Content   string
+	DNidNode  LocalTriState
+	Errors    string
+}
+
+// OrderedAppInfoItem - the struct for key/value
+type OrderedAppInfoItem struct {
+	Key   string       `json:"key"`
+	Value AppInfoItems `json:"value"`
+}
+
+// AppTrackerInfo - the struct for the App and cluster info per node
+type AppTrackerInfo struct {
+	Hostname    string               `json:"hostname"`
+	CollectTime string               `json:"collectTime"`
+	AppInfo     []OrderedAppInfoItem `json:"appInfo"`
+}
+
+// GetApplicationInfo - get Cluster and App related microservices key info from their publications.
+// this library function by given the App-UUID, it can be empty, and returns the node hostname
+// and an array of each of the publication items for debugging purpose.
+// This function can be called from 'zedkube' HTTP handler function on the live node, or
+// it can be called from offline for processing the same microservice publications in eve-info data
+// of the 'collect-info'. The publication directory locations are passed in to handle the various usages.
+func GetApplicationInfo(rootRun, persistStatus, persistKubelog, AppUUID string) AppTrackerInfo {
+	var appInfo []OrderedAppInfoItem
+	var err error
+
+	// 1) Get the Node Info
+	structName := "zedagent-EdgeNodeInfo"
+	nodeInfo := &types.EdgeNodeInfo{}
+	if structName, err = readJSONFile(persistStatus, structName, "global", nodeInfo); err != nil {
+		return addHostnameAppInfo("unknown", appendFailedItem(appInfo, structName, "global", err))
+	}
+	ai := AppInfoItems{ // EdgeNodeInfo
+		UUID:    nodeInfo.DeviceID.String(),
+		Name:    nodeInfo.DeviceName,
+		Exist:   true,
+		Content: fmt.Sprintf("Enterprise: %s, Project: %s", nodeInfo.EnterpriseName, nodeInfo.ProjectName),
+	}
+	appInfo = append(appInfo, OrderedAppInfoItem{Key: structName, Value: ai})
+
+	hostname := nodeInfo.DeviceName
+
+	// 2) Get the EdgeNodeClusterStatus, may not exist
+	structName = "zedkube-EdgeNodeClusterStatus"
+	clusterStatus := &types.EdgeNodeClusterStatus{}
+	if structName, err = readJSONFile(rootRun, structName, "global", clusterStatus); err != nil {
+		appInfo = appendFailedItem(appInfo, structName, "global", err)
+	} else {
+		ClusterIPReady := "Cluster IP Not Ready"
+		if clusterStatus.ClusterIPIsReady {
+			ClusterIPReady = "Cluster IP Ready"
+		}
+		ai := AppInfoItems{ // EdgeNodeClusterStatus
+			UUID:    "global",
+			Name:    clusterStatus.ClusterInterface,
+			Exist:   true,
+			State:   ClusterIPReady,
+			Content: fmt.Sprintf("ClusterIP %v, JoinIP %v, Bootstrap %v", clusterStatus.ClusterIPPrefix, clusterStatus.JoinServerIP, clusterStatus.BootstrapNode),
+		}
+		appInfo = append(appInfo, OrderedAppInfoItem{Key: structName, Value: ai})
+	}
+
+	// 3) Get the KubeLeaseInfo from zedkuube, it may not exist
+	var structName2 string
+	structName = "zedkube-KubeLeaseInfo"
+	leaseInfo := &types.KubeLeaseInfo{}
```

Comment:
This stuct is being renamed in one of your other PRs. Which one should we complete and merge first?

Reply:
we can merge this one first. The other one may need to resolve before merge.

Reply:
Ok